File: web-client/src/app/core/services/course.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ICourse } from 'src/app/Models/Course';
import { QueryParams } from '../types/QueryParams';
import { buildQueryParams } from '../helpers/buildQueryParams';

@Injectable({
  providedIn: 'root',
})
export class CourseService {
  constructor(private _http: HttpClient) {}

  getCourses({
    perPage,
    prev,
  }: QueryParams): Observable<ICourse[]> {
    const queryParams = buildQueryParams({
      perPage: perPage,
      prev: prev,
    });
    const courses = this._http.get<ICourse[]>(
      `/api/Course?${queryParams}`
    );
    return courses;
  }

  getCourse(slug_or_id: string): Observable<ICourse> {
    const course = this._http.get<ICourse>(
      `/api/Course/${slug_or_id}`
    );
    return course;
  }

  saveCourse(course: ICourse): Observable<ICourse> {
    return this._http.post<ICourse>('/api/Course', course);
  }
}

File: api/Services/Implementations/CourseService.cs
using api.Models;
using api.Repositories.Interfaces;
using api.Services.Interfaces;

namespace api.Services.Implementations;
public class CourseService : ICourseService
{
    private readonly ICourseRepository _courseRepository;

    public CourseService(ICourseRepository courseRepository)
    {
        _courseRepository = courseRepository;
    }

    public async Task<Course> GetCourse(int id)
    {
        return await _courseRepository.GetCourse(id);
    }

    public async Task<Course> GetCourse(string slug)
    {
        return await _courseRepository.GetCourse(slug);
    }

    public async Task<IEnumerable<Course>> GetCourses(int prev, int perPage, string? searchTerm)
    {
        return await _courseRepository.GetCourses(prev, perPage, searchTerm);
    }

    public async Task SaveCourse(Course course)
    {
        await _courseRepository.SaveCourse(course);
    }
}
